doctr/models/recognition/sar/tensorflow.py

```diff
@@ -124,39 +146,51 @@ def call(
-        )
-        # run first step of lstm
-        # holistic: shape (N, rnn_units)
-        _, states = self.lstm_decoder(holistic, states, **kwargs)
-        # Initialize with the index of virtual START symbol (placed after <eos> so that the one-hot is only zeros)
-        symbol = tf.fill(features.shape[0], self.vocab_size + 1)
-        logits_list = []
-        if kwargs.get('training') and gt is None:
-            raise ValueError('Need to provide labels during training for teacher forcing')
-        for t in range(self.max_length + 1):  # keep 1 step for <eos>
-            # one-hot symbol with depth vocab_size + 1
-            # embeded_symbol: shape (N, embedding_units)
-            embeded_symbol = self.embed(tf.one_hot(symbol, depth=self.vocab_size + 1), **kwargs)
-            logits, states = self.lstm_decoder(embeded_symbol, states, **kwargs)
-            glimpse = self.attention_module(
-                features, tf.expand_dims(tf.expand_dims(logits, axis=1), axis=1), **kwargs,
-            )
-            # logits: shape (N, rnn_units), glimpse: shape (N, 1)
-            logits = tf.concat([logits, glimpse], axis=-1)
-            # shape (N, rnn_units + 1) -> (N, vocab_size + 1)
-            logits = self.output_dense(logits, **kwargs)
-            # update symbol with predicted logits for t+1 step
-            if kwargs.get('training'):
-                symbol = gt[:, t]  # type: ignore[index]
+        if gt is not None:
+            gt_embedding = self.embed_tgt(gt)
+
+        logits_list: List[tf.Tensor] = []
+
+        for t in range(self.max_length + 1):  # 32
+            if t == 0:
+                # step to init the first states of the LSTMCell
+                states = self.lstm_cells.get_initial_state(
+                    inputs=None, batch_size=features.shape[0], dtype=features.dtype
+                )
+                prev_symbol = holistic
+            elif t == 1:
+                # step to init a 'blank' sequence of length vocab_size + 1 filled with zeros
+                # (N, vocab_size + 1) --> (N, embedding_units)
+                prev_symbol = tf.zeros([features.shape[0], self.vocab_size + 1])
+                prev_symbol = self.embed(prev_symbol)
             else:
-                symbol = tf.argmax(logits, axis=-1)
-            logits_list.append(logits)
-        outputs = tf.stack(logits_list, axis=1)  # shape (N, max_length + 1, vocab_size + 1)
-
-        return outputs
+                if gt is not None:
+                    #(N, embedding_units) -2 because of <bos> and <eos> (same)
+                    prev_symbol = self.embed(gt_embedding[:, t - 2])
+                else:
+                    # -1 to start at timestep where prev_symbol was initialized
+                    index = tf.argmax(logits_list[t - 1], axis=-1)
+                    # (1, embedding_units)
+                    # TODO: translate me :) prev_symbol = prev_symbol.scatter_(1, index.unsqueeze(1), 1)
```

Comment:
@frgfm 
How can we do this in TF ? 
We need to update the prev_symbol last dimension at the given index with 1

Reply:
ok got it